### PR TITLE
feat: Add support for Infracost Cloud based usage estimates

### PR DESCRIPTION
**Pull request by tim775**

### CODE REVIEW

File: internal/schema/core_resource.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	var res *Resource
+	if rb.CoreResource != nil {
+		u := rb.ResourceData.UsageData
+		if fetchedUsage != nil {
+			if u == nil {
+				u = fetchedUsage
+			} else {
+				u = u.Merge(fetchedUsage)
+			}
+		}
```
Comment (hugorut) [inline internal/schema/core_resource.go:43]: I'd probs just move all this logic inside u.Merge, i.e.:

```
func (u *UsageData) Merge(other *UsageData) *UsageData {
    if u == nil {
       return other
    }
    
    if other == nil {
        return nil
    }
    
    ...
}
```

**Review by hugorut — COMMENTED** [review]
Nice, Thanks for the swell code docs - they made it much easier to grasp ❤️ 

So I think we're almost there with this, however I'm still feeling a little iffy about these ` ResourceBuilder` structs.

I think maybe my feelings would be quelled if we turned the `RFunc` to return a `ResourceBuilder` struct, but instead renaming it to `PartialResource` or similar.

This would mean we don't need to itroduce a separate `CoreRFunc` into the resources and files like `parser.go` could be simplified.

So following this approach `DynamoDbTable` would go to:

```go
func NewDynamoDBTableResource(d *schema.ResourceData) *schema.PartialResource {
	...
	
	return &schema.PartialResource{
	     ResourceData: d,
	     CoreResource: &aws.DynamoDBTable{...}
	}
}
```

and old resources would just be:

```go
func NewAPIGatewayStage(d *schema.ResourceData, u *schema.UsageData) *schema.PartialResource {
	r := &aws.APIGatewayStage{
		...
	}

	r.PopulateUsage(u)
	return { 
	    ResourceData: d,
	    Resource: r.BuildResource()
	}
}
```

What's your thoughts?

File: cmd/infracost/run.go [hunk 2]
```diff
@@ -539,6 +543,73 @@ func (r *parallelRunner) runProjectConfig(ctx *config.ProjectContext) (*projectO
...
+}
+
+func (r *parallelRunner) populateActualCosts(projects []*schema.Project) {
+	if r.runCtx.Config.UsageAPIEndpoint != "" {
+		spinnerOpts := ui.SpinnerOptions{
+			EnableLogging: r.runCtx.Config.IsLogging(),
+			NoColor:       r.runCtx.Config.NoColor,
+			Indent:        "  ",
+		}
+		spinner := ui.NewSpinner("Retrieving actual costs for 3 resources from Infracost Cloud", spinnerOpts)
```
[the remark this answers is not in the document]
Comment (tim775) [reply cmd/infracost/run.go:599]: Oops that was prototyping with Ali, I didn't mean to actually commit that.

**Comment by tim775** [comment]
PartialResource is a better name so I will change it to that.  

My main concern (and really the reason for separate `RFunc` and `CoreRFunc` is that I want to delay changing every provider resource.  If i'm going to change the return type for everything it seems like I might as well bite the bullet and refactor everything to return CoreResources.  It just seems like that's a lot more change than I want to deal with now, but on the other hand having two types of resources is confusing and sucky....

I'll due the PartialResource rename and the other feedback items then we can revist it.

**Comment by hugorut** [comment]
> I'll due the PartialResource rename and the other feedback items then we can revist it.

Cool sounds good - I think it makes sense to not change everything for the mo 👍

**Review by hugorut — APPROVED** [review]
🚀

**Pull request merged**